packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -209,6 +209,71 @@ export class LoaderContainerTracker implements IOpProcessingController {
         debugWait("Synchronized");
     }
 
+    /**
+     * Port of {@link @fluidframework/container-loader#waitContainerToCatchUp} for compatibility with old containers.
+     * Waits until container connects to delta storage and gets up-to-date
+     * Useful when resolving URIs and hitting 404, due to container being loaded from (stale) snapshot and not being
+     * up to date. Host may chose to wait in such case and retry resolving URI.
+     * Warning: Will wait infinitely for connection to establish if there is no connection.
+     * May result in deadlock if Container.disconnect() is called and never switched back to auto-reconnect.
+     * @returns true: container is up to date, it processed all the ops that were know at the time of first connection
+     *          false: storage does not provide indication of how far the client is. Container processed
+     *          all the ops known to it, but it maybe still behind.
+     * @throws an error beginning with if the container is closed before it catches up.
+     */
+    public async waitContainerToCatchUp(container: IContainer): Promise<boolean> {
+        return new Promise<boolean>((resolve, reject) => {
+            // Make sure we stop waiting if container is closed.
+            if (container.closed) {
+                reject(new Error("Container is closed"));
+            }
+            const deltaManager = container.deltaManager;
+
+            const closedCallback = (err?: Error | undefined) => {
+                (container as Container).off("closed", closedCallback);
+                reject(new Error("Container is closed"));
+            };
+            (container as Container).on("closed", closedCallback);
+
+            const waitForOps = () => {
+                assert(container.connectionState !== ConnectionState.Disconnected,
+                    0x0cd /* "Container disconnected while waiting for ops!" */);
```

Comment:
You might get forced into generating new error codes, I don't think our policy check likes duplicates.  In that case, just put the string error here for now and let the tool generate it on next release.